### ISSUE TITLE
Media Manager and different root folders for backend users

### ISSUE THREAD

**Issue by Rike-cz** [octobercms/october#4530]
Currently all backend users use Media Manager from predefined root. There is no way how to asocciate user with different root folder, everything is hard-coded. I think many applications of CMS want to have different media root for each backend user, because administrator do not want that users can affect folders and files of other users. But it looks that implementation of this feature would be very difficult, because Media Manager is very much tied to the one folder. But, is there anyone who is able to make this feature? :-)

**Comment by Samuell1** [comment]
This can be possible but only with fully reworking media manager to use database. If you want something like this is better to make custom plugin where people can upload files.

**Comment by Rike-cz** [comment]
I meant only simple solution like:
- root folder /storage/app/media (as now),
- backend user id 1 has root folder: /storage/app/media/1,
- backend user id 2 has root folder: /storage/app/media/2,
etc.

For realize that I need only some event to hook like:
`Event::listen('backend.mediaManager.init', function($mediaManager) {
    $mediaManager->root = ...;
});`

**Comment by alxy** [comment]
You can already do this, as the root folder is a config option. You would just need to override this config option. I know that someone has already sucessfully dont it.

**Comment by Rike-cz** [comment]
@alxy: I tried this monkey patch before (class Plugin):
```
public function register()
{
    \Backend\Models\User::extend(function ($model) {
        $model->bindEvent('model.afterFetch', function() use ($model) {
            \Config::set('cms.storage.media.folder', 'media/' . $model->id);
            \Config::set('cms.storage.media.path', '/storage/app/media/' . $model->id);
        });
    });
}
```
But it causes more troubles with mediafinder form widget and elsewhere in frontend with media filter (when backend user is logged in). This is not good solution.

**Comment by alxy** [comment]
Yeah, this isnt how it worked. Try it in boot and do it in ``Backend\Classes\Controller::extend()``. You should be able to get the current backend user using ``BackendAuth::getUser()`` - no need to hook into the models afterFetch event. I currently do not have an isntallation to test this, but it was something like this.

**Comment by Rike-cz** [comment]
@alxy: It works with Media Manager and controller, but makes troubles with paths for Media finder form widget. Rewriting config on the fly is not really good solution.

**Comment by LukeTowers** [comment]
@Rike-cz rewriting config on the fly is a fine solution, I use it in a number of my plugins / projects. This is something that can be implemented as a plugin with enough work so I'm closing the issue. If you're having trouble making it as a plugin I can whip one up for $150 USD (dev cost, plugin would be released for $5).

**Comment by Rike-cz** [comment]
@LukeTowers I need to integrate this to complicated behavior, but once I have config rewriten, cannot give it back. Best for me would be to have only Media Manager with differerent paths. This works for that good finally:
```
\Backend\Widgets\MediaManager::extend(function ($instance) {
    \Config::set('cms.storage.media.folder', ...);
    \Config::set('cms.storage.media.path', ...);
});
```
But when Media Manager window is closed, I need to set this config to default values, because backend user can see images of another users, he cannot edit/delete them.
That 150 USD is for plugin solution, but not permanent CMS feature, right?

**Comment by LukeTowers** [comment]
@Rike-cz that's correct. 

> But when Media Manager window is closed, I need to set this config to default values, because backend user can see images of another users, he cannot edit/delete them.

Are you saying that you're wanting it to allow backend users to see all of the files on the system, even those added by other users but only allow uploads to a user-specific folder? Why? That's significantly more complex to build.

**Comment by Rike-cz** [comment]
@LukeTowers I simplified my problems to previously mentioned issue. But, what I exactly want is "to lock" backend user write permission in his own folder instead of root folder. Because of security. I do not want to give every user write permission to every file in system, because one user can destroy (delete) whole storage. And yes, best solution would be to let him read permission to whole storage and write permission to his own folder. That's ideal. Because in my use case backend user cannot use foreign images via backend form (mediafinder etc.), I simplified this to whole Media Manager/Browser, but images of different users he can see in other managers, for example like thumbs in relation manager. So rewrite config file forever is not solution for that situation.
